### PR TITLE
[Opt] optimize kernels of find_or_insert, insert_or_assign, hybrid mode.

### PR DESCRIPTION
**Pull request by jiashuy**
- Optimize find_or_insert, insert_or_assign on HBM mode using pipeline and thread parallelism.
- Optimize all Hybrid kernels using digests based filter.
- Others:
  - Fix the value diff error on hybrid mode
     - Enable values check on hybrid mode.
  - Support Sm70 for find_or_insert, insert_or_assign when selecting kernels.
  - Adaptation to new evict strategies.
  - Update the benckmark result at ReadMe.md.

### CODE REVIEW

File: include/merlin/core_kernels/find_or_insert.cuh [hunk 1]
```diff
@@ -21,6 +21,1253 @@
...
+        auto min_score_key = BUCKET::keys(bucket_keys_ptr, min_pos_global);
+        auto expected_key =
+            min_score_key->load(cuda::std::memory_order_relaxed);
+        if (expected_key != static_cast<K>(LOCKED_KEY) &&
+            expected_key != static_cast<K>(EMPTY_KEY)) {
+          auto min_score_ptr =
+              BUCKET::scores(bucket_keys_ptr, BUCKET_SIZE, min_pos_global);
+          bool result = min_score_key->compare_exchange_strong(
+              expected_key, static_cast<K>(LOCKED_KEY),
+              cuda::std::memory_order_acquire, cuda::std::memory_order_relaxed);
```
Comment (rhdong) [inline include/merlin/core_kernels/find_or_insert.cuh:955]: Theoretically, `memory_order_acq_rel` should be more property for the success branch.

File: include/merlin/core_kernels/find_or_insert.cuh [hunk 2]
```diff
@@ -21,6 +21,1253 @@
...
+
+  static void select_kernel(Params& params, cudaStream_t& stream) {
+    const uint32_t total_value_size =
+        static_cast<uint32_t>(params.dim * sizeof(V));
+
+    auto launch_TLPv1 = [&]() {
+      if (total_value_size % sizeof(byte16) == 0) {
+        using VecV = byte16;
+        Launch_TLPv1_FindOrInsert<K, V, S, VecV>::launch_kernel(params,
+                                                                  stream);
```
Comment (rhdong) [inline include/merlin/core_kernels/find_or_insert.cuh]: The code hasn't been formatted.

File: tests/find_or_insert_test.cc.cu [hunk 3]
```diff
@@ -326,31 +326,30 @@ void test_basic_when_full(size_t max_hbm_for_vectors) {
...
   CUDA_CHECK(cudaMalloc(&d_found, KEY_NUM * sizeof(bool)));
 
   CUDA_CHECK(
       cudaMemcpy(d_keys, h_keys, KEY_NUM * sizeof(K), cudaMemcpyHostToDevice));
   CUDA_CHECK(cudaMemcpy(d_scores, h_scores, KEY_NUM * sizeof(S),
                         cudaMemcpyHostToDevice));
 
-  CUDA_CHECK(cudaMemset(d_vectors, 1, KEY_NUM * sizeof(V) * options.dim));
+  CUDA_CHECK(cudaMemcpy(d_vectors, h_vectors,
+                          KEY_NUM * sizeof(V) * options.dim,
```
Comment (rhdong) [inline tests/find_or_insert_test.cc.cu]: No formatted.

File: include/merlin/core_kernels/find_or_insert.cuh [hunk 4]
```diff
@@ -21,6 +21,1253 @@
...
+        for (int j = 0; j < Load_LEN_S; j += 1) {
+          S temp_score = temp_scores[j];
+          if (temp_score < min_score) {
+            min_score = temp_score;
+            min_pos = i + k + j;
+          }
+        }
+      }
+    }
+    score = scores == nullptr ? device_nano<S>() : score;
```
Comment (rhdong) [inline include/merlin/core_kernels/find_or_insert.cuh]: Please check the code of multi-strategy support and refactor them by referring to the[ PR#152](https://github.com/NVIDIA-Merlin/HierarchicalKV/pull/152/) like https://github.com/NVIDIA-Merlin/HierarchicalKV/pull/152/files#diff-271acb18bcae4b5c9972e32f590ac5f2404ee62e7079c1b343c8288161cf8870R106

Comment (jiashuy) [reply include/merlin/core_kernels/find_or_insert.cuh]: Yes, I support the new evict strategy in the second commit. And I will merge these three commit to one to facilite review.

File: tests/find_or_insert_test.cc.cu (hunk 3 above)
Comment (jiashuy) [reply tests/find_or_insert_test.cc.cu]: Done by the second commit.

File: include/merlin/core_kernels/find_or_insert.cuh (hunk 1 above)
Comment (jiashuy) [reply include/merlin/core_kernels/find_or_insert.cuh:955]: I think memory_order_acquire is enough.
As we can gurantee that the modificatio to bucket can't be reorderd before the compare_exchange_strong()

**Pull request merged**